services/device-registry-base/src/test/java/org/eclipse/hono/service/management/device/AutoProvisioningTest.java

```diff
@@ -0,0 +1,215 @@
+        final CertificateFactory factory = CertificateFactory.getInstance("X.509");
+
+        CERT = (X509Certificate) factory.generateCertificate(new FileInputStream(ssc.certificatePath()));
+        SUBJECT_DN = CERT.getSubjectX500Principal().getName(X500Principal.RFC2253);
+    }
+
+    /**
+     * Initializes common fixture.
+     */
+    @BeforeEach
+    public void init() {
+        tenantId = UUID.randomUUID().toString();
+        deviceId = UUID.randomUUID().toString();
+        deviceManagementService = mock(DeviceManagementService.class);
+        credentialsManagementService = mock(CredentialsManagementService.class);
+        tenantInformationService = mock(TenantInformationService.class);
+    }
+
+    /**
+     * Verifies that auto-provisioning of a device succeeds.
+     *
+     * @param ctx The vert.x test context.
+     * @throws CertificateEncodingException if the certificate cannot be encoded.
+     */
+    @Test
+    public void testProvisionDeviceSucceeds(final VertxTestContext ctx) throws CertificateEncodingException {
+        // GIVEN a tenant CA with auto-provisioning enabled
+        final TenantObject tenantObject = TenantObject.from(tenantId).addTrustAnchor(CERT.getPublicKey(),
+                CERT.getSubjectX500Principal(), true);
+        final JsonObject clientContext = new JsonObject().put(CredentialsConstants.FIELD_CLIENT_CERT,
+                CERT.getEncoded());
+
+        when(tenantInformationService.getTenant(eq(tenantId), any()))
+                .thenReturn(Future.succeededFuture(TenantResult.from(HttpURLConnection.HTTP_OK, tenantObject)));
+        when(deviceManagementService.createDevice(eq(tenantObject.getTenantId()), any(), any(), any()))
+                .thenReturn(Future.succeededFuture(OperationResult.ok(HttpURLConnection.HTTP_OK, Id.of(deviceId),
+                        Optional.empty(), Optional.empty())));
+        when(credentialsManagementService.updateCredentials(eq(tenantObject.getTenantId()), eq(deviceId), any(), any(),
+                any())).thenReturn(Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_CREATED)));
+
+        // WHEN provisioning a device from a certificate
+        AutoProvisioning
+                .provisionIfEnabled(tenantId, SUBJECT_DN, clientContext, credentialsManagementService,
+                        deviceManagementService, tenantInformationService, NoopSpan.INSTANCE)
+                .onComplete(ctx.succeeding(result -> {
+                    ctx.verify(() -> {
+                        // THEN the device is registered, credentials are set and credentials information is returned
+                        verify(deviceManagementService).createDevice(eq(tenantObject.getTenantId()), any(),
+                                any(), any());
+                        verify(credentialsManagementService).updateCredentials(eq(tenantObject.getTenantId()),
+                                eq(deviceId), any(), any(), any());
```

Comment:
Since only one device is being auto-provisioned in each test, I didn't change those assertions that I moved from the tests from `AutoProvisioningEnabledDeviceBackendTest`. It's good to be more specific and I will update it.

Reply:
Since only one device is being auto-provisioned in each test, I didn't change those assertions that I moved from the tests from `AutoProvisioningEnabledDeviceBackendTest`. It's good to be more specific and I will update it.

Reply:
I have pushed a commit adding more assertions.